### ISSUE TITLE
Build fail due to timeout (not due to dependencies)

### ISSUE THREAD

**Issue by rgerkin** [readthedocs/readthedocs.org#6691]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/neuronunit
* Build URL (if applicable): https://readthedocs.org/projects/neuronunit/builds/10482732/
* Read the Docs username (if applicable): rgerkin

## Expected Result

A build that succeeds.

## Actual Result

A build timeout.  I even made a new requirements file that stripped all dependencies, but the build times out after processing `conf.py` (which it reports as complete after `Command time: 0s Return: 0`).  Total command time for all steps is only 35 seconds, so it appears to hang for ~8.5 minutes before timing out.  Ctrl-F shows no "error" or "exception" anywhere during the build up to that point.

**Comment by humitos** [comment]
Hi @rgerkin! This happen when trying to build the "htmlzip" file (all the documentation in just one file) which takes _a lot of resources_.

I recommend you to use our Config File V2 (https://docs.readthedocs.io/en/stable/config-file/v2.html) which by default disables this output format. That should be enough to make your builds to work.

**Comment by rgerkin** [comment]
@humitos Thanks for the suggestion, but it is still timing out.
Build: https://readthedocs.org/projects/neuronunit/builds/10496600
Current config file: https://github.com/scidash/neuronunit/blob/docs/.readthedocs.yml

**Comment by humitos** [comment]
I increased the building time allowed. Although, I think most of the time is used to document the API (since there is no too many documentation pages), so I suggest you to take a look at https://sphinx-autoapi.readthedocs.io/ (instead of `sphinx.ext.autodoc`) which will help you to improve the building time of the module API.

Let me know if this works.

**Comment by humitos** [comment]
Hrm... It failed even increasing the time to 1500s: https://readthedocs.org/projects/neuronunit/builds/10496823/

**Comment by rgerkin** [comment]
@humitos Yes, I saw that.  We are going to try to rebuild `conf.py` from scratch and see what we can make work.

**Comment by rgerkin** [comment]
@humitos Well it certainly [builds successfully](https://readthedocs.org/projects/neuronunit/builds/10498077/) when we remove `sphinx.ext.autodoc`, but of course then it is missing the API documentation.  We will look into `sphinx-autoapi`, but it is strange that autodoc times out on the readthedocs servers, because it only takes about 10 seconds more (relative to not including autodoc) on a regular desktop, and I doubt the readthedocs VMs are 100x slower than that.  Is there any other debug information that we could access?

**Comment by rgerkin** [comment]
@humitos Well it certainly [builds successfully](https://readthedocs.org/projects/neuronunit/builds/10498077/) when we remove `sphinx.ext.autodoc`, but of course then it is missing the API documentation.  We will look into `sphinx-autoapi`, but it is strange that autodoc times out on the readthedocs servers, because it only takes about 10 seconds more (relative to not including autodoc) on a regular desktop, and I doubt the readthedocs VMs are 100x slower than that.  Is there any other debug information that we could access?

**Comment by humitos** [comment]
Wow, 71 seconds without the `autodoc`! I don't have any advice from the top of my head right now about what it could be considering than you said locally it takes only 10s more.

**Comment by ChihweiLHBird** [comment]
> Wow, 71 seconds without the `autodoc`! I don't have any advice from the top of my head right now about what it could be considering than you said locally it takes only 10s more.

May we know which command that readthedoc was sucks on? Maybe we can try the same thing locally.

**Comment by rgerkin** [comment]
@humitos Is there something that executes between the listed commands on the build page, but which isn't shown?

**Comment by humitos** [comment]
> May we know which command that readthedoc was sucks on? Maybe we can try the same thing locally.

The command that follow after `cat conf.py` is this one:

```
 python /home/docs/checkouts/readthedocs.org/user_builds/neuronunit/envs/docs/bin/sphinx-build -T -b readthedocs -d _build/doctrees-readthedocs -D language=en . _build/html 
```

> @humitos Is there something that executes between the listed commands on the build page, but which isn't shown?

No.

Let me know if that helps.

**Comment by humitos** [comment]
I tried building it locally to debug this and double check if the problem is under Read the Docs or not. The `docs` branch passes without problems, so I suppose that you made some changes there to make it work. What would be the branch that I should use to test locally the timeout issue you had on RTD?

**Comment by rgerkin** [comment]
@humitos That's right, I removed autodoc from the `docs` branch so that it would build (and changed the root from `index.rst` to `contents.rst` which is required later in the build).  I just created a new branch `docs-fail` which retains autodoc, so you can use that to test.  I also just tested `docs` and `docs-fail` again [here](https://readthedocs.org/projects/neuronunit/builds/10515566) and [here](https://readthedocs.org/projects/neuronunit/builds/10515573/), respectively, to verify the behavior, which shows the former building successfully and the latter timing out.

**Comment by humitos** [comment]
I cloned the repo and try building it locally. I got different errors when running `sphinx-build` around modules not found even after installing the package `neurounit` itself.

I could help you more to debug this issue on RTD once the repository has a branch that I can test by myself.

**Comment by rgerkin** [comment]
@humitos Do you mean that you want a branch that is guaranteed to work locally for you, and then you can test it on RTD?  It works locally for me, but perhaps not in a clean environment.  I'll make whatever changes are necessary to get it running from a clean environment.

**Comment by humitos** [comment]
I'm basically,

1. clonning the repo
2. switch to `docs-fail` branch
3. creating a virtualenv
4. installing sphinx
5. installing your package (pip install .)
6. running sphinx-build

That fails with issues.

If you can share a detailed list of full commands that need to be executed from a clean environment, I may be able to help on this.

**Comment by rgerkin** [comment]
I was able to resolve this.  I'm not 100% sure what the source of the problem was, because I did several things (including reducing the number of warnings, and switching from `autodoc` to `sphinx-apidoc` as you suggested), but I think it was because in one of the modules imported by `autodoc`, there was a system call (using `os.system`) that was not returning control to the build process without user input.  This was in a random unit test module that isn't even part of the normal test suite, so I hadn't noticed it before.  I further added `*unit_test*` to the `exclude` list in `conf.py`, since I don't need my unit tests documented here.  The new build is [here](https://readthedocs.org/projects/neuronunit/builds/10529375/).  Thanks for your help.
